src/cdk-experimental/ui-patterns/behaviors/list-navigation/list-navigation.ts

```diff
@@ -0,0 +1,122 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import {signal, Signal, WritableSignal} from '@angular/core';
+
+/** Represents an item in a collection, such as a listbox option, than can be navigated to. */
+export interface ListNavigationItem {
+  /** Whether an item is disabled. */
+  disabled: Signal<boolean>;
+}
+
+/** Represents the required inputs for a collection that has navigable items. */
+export interface ListNavigationInputs<T extends ListNavigationItem> {
+  /** Whether focus should wrap when navigating. */
+  wrap: Signal<boolean>;
+
+  /** The items in the list. */
+  items: Signal<T[]>;
+
+  /** Whether disabled items in the list should be skipped when navigating. */
+  skipDisabled: Signal<boolean>;
+
+  /** The current index that has been navigated to. */
+  activeIndex: WritableSignal<number>;
+
+  /** Whether the list is vertically or horizontally oriented. */
+  orientation: Signal<'vertical' | 'horizontal'>;
+
+  /** The direction that text is read based on the users locale. */
+  textDirection: Signal<'rtl' | 'ltr'>;
+}
+
+/** Controls navigation for a list of items. */
+export class ListNavigation<T extends ListNavigationItem> {
+  /** The last index that was active. */
+  prevActiveIndex = signal(0);
+
+  constructor(readonly inputs: ListNavigationInputs<T>) {
+    this.prevActiveIndex.set(inputs.activeIndex());
+  }
+
+  /** Navigates to the given item. */
+  goto(item: T) {
+    if (this.isFocusable(item)) {
+      this.prevActiveIndex.set(this.inputs.activeIndex());
+      const index = this.inputs.items().indexOf(item);
+      this.inputs.activeIndex.set(index);
+    }
+  }
+
+  /** Navigates to the next item in the list. */
+  next() {
+    const items = this.inputs.items();
+
+    for (let i = this.inputs.activeIndex() + 1; i < items.length; i++) {
+      if (this.isFocusable(items[i])) {
+        this.goto(items[i]);
+        return;
+      }
+    }
+
+    if (this.inputs.wrap()) {
+      for (let i = 0; i <= this.inputs.activeIndex(); i++) {
+        if (this.isFocusable(items[i])) {
+          this.goto(items[i]);
+          return;
+        }
+      }
+    }
+  }
+
+  /** Navigates to the previous item in the list. */
+  prev() {
+    const items = this.inputs.items();
+
+    for (let i = this.inputs.activeIndex() - 1; i >= 0; i--) {
+      if (this.isFocusable(items[i])) {
+        this.goto(items[i]);
+        return;
+      }
+    }
+
+    if (this.inputs.wrap()) {
+      for (let i = items.length - 1; i >= this.inputs.activeIndex(); i--) {
+        if (this.isFocusable(items[i])) {
+          this.goto(items[i]);
+          return;
+        }
+      }
+    }
+  }
+
```

Comment:
I don't love the duplication here for going the other direction- I think it's valuable to keep the code size down even if the code itself is less obvious